File: MosaicMaker/MosaicMaker.ts
/**
 * MosaicMaker.ts
 *
 * Creates a mosaic image png
 * usage: ts-node MosaicMaker.ts TILE_SIZE "input url/filepath" "output png file"
 * example: ts-node MosaicMaker.ts 20 "https://raw.githubusercontent.com/Maxoplata/misc/master/MosaicMaker/_readmeAssets/sampleInput.jpg" "./mosaic.png"
 * example: ts-node MosaicMaker.ts 20 "./sampleInput.jpg" "./mosaic.png"
 *
 * @author Maxamilian Demian
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/misc
 */

import * as Jimp from 'jimp';

(async () => {
	if (process.argv.length != 5) {
		throw new Error('Invalid argument count');
	}
	const tileSize: number = parseInt(process.argv[2]);
	const inputFile: string = process.argv[3];
	const outputFile: string = process.argv[4];

	// validate tile size
	if (tileSize < 2) {
		throw new Error('Invalid tile size (minimum 2)');
	}

	try {
		// load input file
		let imgOrig: Jimp = await Jimp.read(inputFile);

		// get pixel colors from original image
		const widthOrig: number = imgOrig.bitmap.width;
		const heightOrig: number = imgOrig.bitmap.height;
		const pixelColors: number[][] = [];

		for (let x: number = 0; x < widthOrig; x++) {
			pixelColors[x] = [];

			for (let y: number = 0; y < heightOrig; y++) {
				pixelColors[x][y] = imgOrig.getPixelColor(x, y);
			}
		}

		// convert original image to tile
		imgOrig.resize(tileSize, tileSize);

		// create new image
		const imgNew: Jimp = new Jimp(widthOrig * tileSize, heightOrig * tileSize);

		// add tiles to new image
		for (let x: number = 0; x < widthOrig; x++) {
			for (let y: number = 0; y < heightOrig; y++) {
				// add original image tile
				imgNew.composite(imgOrig, x * tileSize, y * tileSize, {
					mode: Jimp.BLEND_SOURCE_OVER,
					opacitySource: 1,
					opacityDest: 1,
				})

				// create color tile
				const imgColorTile: Jimp = new Jimp(tileSize, tileSize, pixelColors[x][y]);

				// add translucent color tile
				imgNew.composite(imgColorTile, x * tileSize, y * tileSize, {
					mode: Jimp.BLEND_SOURCE_OVER,
					opacitySource: 0.63,
					opacityDest: 1,
				})
			}
		}

		// output image
		await imgNew.write(outputFile);
	} catch (error) {
		throw new Error(`Failed to create Mosaic: ${error}`)
	}
})();

File: MosaicMaker/MosaicMaker.py
#!/usr/bin/env python
"""
MosaicMaker.py

Creates a mosaic image png
usage: python MosaicMaker.py TILE_SIZE INPUT_TYPE "input url/filepath" "output png file"
example: python MosaicMaker.py 20 "https://raw.githubusercontent.com/Maxoplata/misc/master/MosaicMaker/_readmeAssets/sampleInput.jpg" "./mosaic.png"
example: python MosaicMaker.py 20 "./sampleInput.jpg" "./mosaic.png"

https://www.maxodev.org
https://github.com/Maxoplata/misc
"""

import cStringIO
import os
import sys
import urllib2
from PIL import Image

__author__ = "Maxamilian Demian"
__email__ = "max@maxdemian.com"

# if we have arguments passed to the script
if len(sys.argv) != 4:
	raise Exception('Invalid argument count')

# vars
tileSize = int(sys.argv[1])
inputFile = sys.argv[2]
outputFile = sys.argv[3]

# validate the tile size
if tileSize < 2:
	raise Exception('Invalid tile size (minimum 2)')

# validate input file
urlData = None

if not os.path.exists(inputFile):
	try:
		urlData = urllib2.urlopen(inputFile).read()
	except:
		raise Exception('File does not exist')

try:
	# validate image file
	imgOrig = Image.open(inputFile) if urlData == None else Image.open(cStringIO.StringIO(urlData))

	# get width/height of image
	widthOrig = imgOrig.size[0]
	heightOrig = imgOrig.size[1]

	# create tile of image
	imgOrigResized = imgOrig.resize((tileSize, tileSize))

	# create new image
	imgNew = Image.new('RGB', ((widthOrig * tileSize), (heightOrig * tileSize)))

	# load image pixel data
	imgOrig = imgOrig.convert('RGB')
	pixels = imgOrig.load()

	# iterate through original image pixels
	for x in range(0, widthOrig):
		for y in range(0, heightOrig):
			# add tile to new image
			imgNew.paste(imgOrigResized, ((x * tileSize), (y * tileSize)))

			# get pixel color from original image
			rgb = pixels[x, y]
			r = rgb[0]
			g = rgb[1]
			b = rgb[2]

			# add translucent color layer over tile
			imgColorTile = Image.new('RGBA', (tileSize, tileSize), (r, g, b, 127))
			imgNew.paste(imgColorTile, ((x * tileSize), (y * tileSize)), imgColorTile)

	# save file
	imgNew.save(outputFile, 'PNG')
except Exception as error:
	raise Exception('Failed to create Mosaic: {}'.format(error))
